atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -69,157 +74,191 @@
+    @FunctionalInterface
+    interface LoggingFunction {
+        void log(String fmt, Object... args);
     }
 
-    private static void logTime(String method, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} took {} ms.", method, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    @FunctionalInterface
+    interface TriConsumer<A,B,C> {
+        void accept(A aa, B bb, C cc);
     }
 
-    private static void logTimeAndTable(String method, String tableName, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} took {} ms.",
+    private static void logCellsAndSize(LoggingFunction logger, String method, String tableName, int numCells, long sizeInBytes, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} for {} cells of overall size {} bytes took {} ms.",
+                tableName, numCells, sizeInBytes, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+
+    private static void logTime(LoggingFunction logger, String method, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} took {} ms.", method, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+
+    private static void logTimeAndTable(LoggingFunction logger, String method, String tableName, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} took {} ms.",
                 method, tableName, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableCount(String method, int tableCount, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} for {} tables took {} ms.",
+    private static void logTimeAndTableCount(LoggingFunction logger, String method, int tableCount, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} for {} tables took {} ms.",
                 method, tableCount, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableRange(String method, String tableName, RangeRequest range, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} with range {} took {} ms.",
+    private static void logTimeAndTableRange(LoggingFunction logger, String method, String tableName, RangeRequest range, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} with range {} took {} ms.",
                 method, tableName, range, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
+    private void maybeLog(Runnable runnable, Predicate<Stopwatch> shouldSlowLog, BiConsumer<LoggingFunction, Stopwatch> logger) {
+        if (log.isTraceEnabled() || slowlogger.isWarnEnabled()) {
+            Stopwatch stopwatch = Stopwatch.createStarted();
+            runnable.run();
+            stopwatch.stop();
+            if (log.isTraceEnabled()) {
+                logger.accept(log::trace, stopwatch);
+            }
+            if (slowlogger.isWarnEnabled() && shouldSlowLog.test(stopwatch)) {
+                logger.accept(slowlogger::warn, stopwatch);
```

Comment:
we probably still want to log in the case of exceptions (especially if things timeout), so we should be doing the logging from finally blocks.